### DISCUSSION TITLE
Adding SlateJS plugins

### DISCUSSION THREAD

**Discussion by frederico-taramaschi** [payloadcms/payload#7769] [Q&A]
How would someone go about adding a plugin (https://github.com/nlulic/slate-table) to the slateJS implementation in Payload (v2)? This does not appear to work the normal way.

**Comment by tyteen4a03** [comment]
Not an answer to your question but Slate is deprecated. Is there a reason you can't switch to Lexical?

**Reply by frederico-taramaschi** [reply]
The Payload CMS docs still point to Slate as being the go to editor rather than Lexical.
> Payload's Lexical rich text editor is currently in beta. It's stable enough to use as you build on Payload, so if you're up for helping us fine-tune it, you should use it. But if you're looking for stability, use Slate instead.

And switching over at a point where the project is 90% finished would be a significant pain.

**Reply by tyteen4a03** [reply]
> The Payload CMS docs still point to Slate as being the go to editor rather than Lexical.
> 
> > Payload's Lexical rich text editor is currently in beta. It's stable enough to use as you build on Payload, so if you're up for helping us fine-tune it, you should use it. But if you're looking for stability, use Slate instead.
> 
> And switching over at a point where the project is 90% finished would be a significant pain.

@AlessioGr might want to fix this!

**Comment by frederico-taramaschi** [comment]
How is something so fundamentally basic not covered or this hard to answer? Boggles the mind honestly.

**Reply by denolfe** [reply]
The ['Building Custom Elements and Leaves'](https://payloadcms.com/docs/rich-text/slate#building-custom-elements-and-leaves)  section of the docs would be the best resource for doing this.

Largely, this is getting your hands dirty with Slate itself and passing those into Payload's rich text field via elements and leaves appropriately.

**Reply by frederico-taramaschi** [reply]
Getting an instance of Slate working with the extra plugin on its own is easy enough and so is adding an Element to the Slate editor in Payload. However the plugins property doesn't seem like it can handle a plugin meant to work like the one I wish to use. Unless I am missing something, which is definitely possible but in any case to way you would "normally" do it doesn't seem feasible?

**Comment by mvdve** [comment]
We have a project which is already in production for more then a year so i can't switch the rich text editor to Lexical so i dived a bit into the documentation of the SlateJS editor. Something like this should work (haven't tested is as i am building a translation button for the editor):

```ts
import {withTable} from "slate-table";

const richTextElements: RichTextElement[] = [
    "h2",
    "h3",
    "h4",
    "h5",
    "h6",
    "link",
    "ol",
    "ul",
    "indent",
    "blockquote",
    {
        name: "instagramPost",
        Button: InstagramPostButton,
        Element: InstagramPostElement,
    },
    {
        name: "table",
        Button: // this is mandatory, so maybe just a dummy component....,
        Element: // this is mandatory, so maybe just a dummy component....,
        plugins: [(editor) => withTable(editor)],
    },
];
```
I did not find a global plugin option in the slateJS config.

**Reply by frederico-taramaschi** [reply]
`plugins: [(editor) => withTable(editor)]`
That is what I expected as well but it is not the case it seems.

For the plugins part anyway the Button component does need something more than just a dummy component seeing how the table plugin works specifically.

**Reply by mvdve** [reply]
Hmm running in to the same issue. The plugin function does not seem to do anything..

**Reply by frederico-taramaschi** [reply]
It's not that it doesn't do *anything* because I *can* change/update the editor instance in the plugins property but it doesn't seem to behave like you would expect and seems like it doesn't handle plugins like the table plugin very well but I could be wrong.

**Reply by mvdve** [reply]
Looking at the [RichText component itself](https://github.com/payloadcms/payload/blob/main/packages/richtext-slate/src/field/RichText.tsx#L230), it looks like there is no option to initialize with custom plugins. You could copy the full component into a new one and add your plugin there. Then override the field component:

```ts
admin: {
    components: {
        Field: RichTextWithPlugin,
    },
},
```

**Reply by berkerdemirer** [reply]
Have you figured this out @frederico-taramaschi @mvdve ? I am having the same troubles adding table support to slate in v2

**Reply by mvdve** [reply]
@berkerdemirer I copied the [full componenten](https://github.com/payloadcms/payload/blob/v2.27.0/packages/richtext-slate/src/field/RichText.tsx) and added some custom functionality. You have to add @ts-ignore here and there, as payload is not developed in strict mode and there are a couple of type issues.
